ec2_elb_facts: Handle AWS throttling when requesting ELB information

bobobox opened the pull request:
##### SUMMARY
Fixes issue where module would fail immediately upon encountering AWS API throttling.
Also makes minor changes to error handling.

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
ec2_elb_facts.py

##### ANSIBLE VERSION
```
ansible 2.2.1.0
  config file =
  configured module search path = Default w/o overrides
```


##### ADDITIONAL INFORMATION
Uses the `AWSRetry.backoff` method both as a decorator and, where more efficient, directly as a wrapper function to  handle AWS API throttling for boto methods invoked in the module. 

Various instances of `BotoServerError` exceptions being caught and handled simply by invoking the  `fail_json` method have been removed, and replaced with a similar try/catch block in the `main` function. I'm not sure that this exception handling was particularly useful in the first place (and I'd be open to removing it completely), but moving it into `main` at least provides pretty much equivalent behavior without interfering with use of the `AWSRetry` decorator, which can handle specific cases of `BotoServerError`.

Also, removes questionable 'catch-all' clause on line 111 which would return an empty dict for tags.

bobobox committed:
Handle throttling, and also move BotoServerError catching to main.

Unclear how useful it is to still be catching BotoServerError exceptions
generally, especially when we are now hopefully handling most instances
of throttling, but kept it as it was already in the code. Moved to main
as catching BotoServerErrors inside individual functions will bypass any
AWSRetry.backoff decorator.

bobobox committed:
Move additional imports to top.

gundalow commented on lib/ansible/modules/cloud/amazon/ec2_elb_facts.py:
```diff
@@ -80,6 +80,9 @@
 
 import traceback
 
+from ansible.module_utils.basic import *
```
Please import what you need rather than `*`

http://docs.ansible.com/ansible/dev_guide/developing_modules_documenting.html#python-imports

gundalow replied:
Ah, I see you've just moved the code to satisfy CI, rather than these been new imports. If you could fix this it would be great.

mjschultz replied:
I fixed them in https://github.com/ansible/ansible/pull/22415 which I can fix up once this is merged in.

bobobox replied:
I merged @mjschultz's fixes to address @gundalow's request.

gundalow replied:
Sounds good

gundalow commented:
If this can get reviewed by the maintainers within the week we can look at getting this into the 2.3 release.

mjschultz commented:
shipit

mjschultz committed:
Use python2.6 formatting

mjschultz committed:
A bit of cleanup on imports

bobobox commented:
I merged @mjschultz's fixes to address @gundalow's request.

s-hertel commented:
shipit

willthames commented on lib/ansible/modules/cloud/amazon/ec2_elb_facts.py:
```diff
@@ -240,24 +239,28 @@ def main():
-    ec2_facts_result = dict(changed=False,
-                            elbs=elb_information.list_elbs())
+    except BotoServerError as err:
+        self.module.fail_json(msg="{}: {}".format(err.error_code, err.error_message),
```
This will fail on python 2.6

bobobox replied:
Thanks. Fixed in 189eb0f42

willthames commented on lib/ansible/modules/cloud/amazon/ec2_elb_facts.py:
```diff
@@ -80,6 +80,14 @@
 import traceback
 
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.ec2 import (
```
Just put this on one line (or two if needed for line lengths)
```
from a.b.c import x, y, z
from a.b.c import u, v, w
```

bobobox replied:
Is there an established Ansible style that says this should be kept on one line? (honest question, I am new to the codebase)
In general I prefer putting things like this into multiple lines as it's a lot clearer later on in diffs when elements are added or removed, and overall just easier (IMO) to read when there are more than a couple items.

willthames replied:
I'd suggest it's an undocumented preference

But it's a fairly standard way of doing things, at least in the amazon modules:

```
$ git grep  'import [^(]*,' | wc -l
75
$ git grep 'import (' | wc -l
5
```

bobobox committed:
Make one more string interpolation Python 2.6 compatible.

bobobox commented:
@gundalow - could you approve your requested changes? Thanks :)

ryansb committed:
Format ELBInfo call

Move declaration of AWSRetry of get_all_load_balancers so it isn't redeclared for every page of data.

ryansb committed:
Remove extra newlines

ryansb commented:
Merged as https://github.com/ansible/ansible/commit/6d8edfc35ff469fbc39c52a0b5ec391aa290ae4a
